cmd/imagec/docker.go

```diff
@@ -181,47 +183,64 @@ func FetchImageBlob(options ImageCOptions, image *ImageWithMeta) error {
 	}()
 
 	// Open the file so that we can use it as a io.Reader for sha256 calculation
-	in, err := os.Open(string(imageFileName))
+	imageFile, err := os.Open(string(imageFileName))
 	if err != nil {
-		return err
+		return diffID, err
 	}
-	defer in.Close()
+	defer imageFile.Close()
 
-	h := sha256.New()
+	// blobSum is the sha of the compressed layer
+	blobSum := sha256.New()
+
+	// diffIDSum is the sha of the uncompressed layer
+	diffIDSum := sha256.New()
+
+	// blobTr is an io.TeeReader that writes bytes to blobSum that it reads from imageFile
+	// see https://golang.org/pkg/io/#TeeReader
+	blobTr := io.TeeReader(imageFile, blobSum)
 
 	progress.Update(po, image.String(), "Verifying Checksum")
-	if _, cerr := io.Copy(h, in); cerr != nil {
-		return cerr
+	tar, err := archive.DecompressStream(blobTr)
+	if err != nil {
+		return diffID, err
 	}
 
-	// Calculate the sum
-	sum := fmt.Sprintf("sha256:%x", h.Sum(nil))
-	if sum != layer {
-		return fmt.Errorf("Failed to validate layer checksum. Expected %s got %s", layer, sum)
+	// Copy bytes from decompressed layer into diffIDSum to calculate diffID
+	if _, cerr := io.Copy(diffIDSum, tar); cerr != nil {
+		return diffID, cerr
 	}
 
+	bs := fmt.Sprintf("sha256:%x", blobSum.Sum(nil))
+	if bs != layer {
```

Comment:
Just curios.  Does this test work?  I couldnt tell if docker hub was doing a sum over the files or a sum over the tar and wasn't sure what to check.